File: back-end/po-service/src/purchase-orders/state-machine.js
const { AppError } = require('../common/errors');

const PO_TRANSITIONS = [
  { from: 'DRAFT',     to: 'SUBMITTED', role: 'BUYER' },
  { from: 'DRAFT',     to: 'CANCELLED', role: 'BUYER' },
  { from: 'SUBMITTED', to: 'APPROVED',  role: 'APPROVER' },
  { from: 'SUBMITTED', to: 'REJECTED',  role: 'APPROVER' },
  { from: 'APPROVED',  to: 'FULFILLED', role: 'BUYER' },
  { from: 'APPROVED',  to: 'CANCELLED', role: 'BUYER' },
  { from: 'REJECTED',  to: 'DRAFT',     role: 'BUYER' },
  { from: 'FULFILLED', to: 'CLOSED',    role: 'BUYER' },
];

function assertTransition(from, to) {
  if (!PO_TRANSITIONS.some(t => t.from === from && t.to === to)) {
    throw new AppError(409, 'INVALID_STATUS_TRANSITION', `Cannot transition from ${from} to ${to}`, { from, to });
  }
}

module.exports = { PO_TRANSITIONS, assertTransition };

File: back-end/supplier-service/app/state_machine.py
from .errors import AppError

# (from, to) -> required role
TRANSITIONS = {
    ("PENDING_APPROVAL", "ACTIVE"): "ADMIN",
    ("PENDING_APPROVAL", "BLACKLISTED"): "ADMIN",
    ("ACTIVE", "INACTIVE"): "ADMIN",
    ("ACTIVE", "BLACKLISTED"): "ADMIN",
    ("INACTIVE", "ACTIVE"): "ADMIN",
    ("INACTIVE", "BLACKLISTED"): "ADMIN",
}

def assert_transition(current: str, target: str):
    if (current, target) not in TRANSITIONS:
        raise AppError(
            409, "INVALID_STATUS_TRANSITION",
            f"Cannot transition from {current} to {target}",
            {"from": current, "to": target},
        )
